eradiate/scenes/biosphere.py

```diff
@@ -436,308 +794,112 @@ def __str__(self):
-                values = line.split()
+            # If a unit is found, try to apply it
+            # Bonus: if a unit field *and* a quantity were found, we convert the
+            # quantity to the unit
+            field_value = d_copy[field]
+            d_copy[field] = ensure_units(field_value, field_units, convert=True)
 
-                if i == 0:
-                    radius = ensure_units(float(values[0].strip()), "meter")
+        size = d_copy.pop("size")
+        if "leaf_definition" in d_copy:
+            leaf_cloud =  LeafCloud.from_rami(**d_copy)
+        else:
+            if d_copy["shape_type"] != "cube":
+                raise ValueError("The HomogeneousDiscreteCanopy must be used"
+                                 "with a cuboid leaf cloud.")
+            leaf_cloud = LeafCloud.from_parameters(**d_copy)
 
-                position = [
-                    float(values[1].strip()),
-                    float(values[2].strip()),
-                    float(values[3].strip()),
-                ]
+        return cls(leaf_cloud_specs=[(leaf_cloud, [[0, 0, 0]])], size=size)
 
-                positions_.append(position)
-                normal = [
-                    float(values[4].strip()),
-                    float(values[5].strip()),
-                    float(values[6].strip()),
-                ]
 
-                orientations_.append(normal)
+@BiosphereFactory.register("floating_spheres_canopy")
+@parse_docs
+@attr.s
+class FloatingSpheresCanopy(DiscreteCanopy):
+    """Generates the canopy for the `floating sphere scenarios<https://rami-benchmark.jrc.ec.europa.eu/HTML/RAMI3/EXPERIMENTS3/HETEROGENEOUS/FLOATING_SPHERES/SOLAR_DOMAIN/DISCRETE/DISCRETE.php>`_
+    in the RAMI benchmark"""
 
-        positions = np.array(positions_) * ureg.m
-        orientations = np.array(orientations_)
+    id = documented(
+        attr.ib(
+            default="floating_spheres_canopy",
+            validator=attr.validators.optional(attr.validators.instance_of(str)),
+        ),
+        doc=get_doc(Canopy, "id", "doc"),
+        type=get_doc(Canopy, "id", "type"),
+        default="floating_spheres_canopy",
+    )
 
-        return cls(
-            leaf_reflectance=leaf_reflectance,
-            leaf_transmittance=leaf_transmittance,
-            leaf_radius=radius,
-            leaf_positions=positions,
-            leaf_orientations=orientations,
-            size=size,
-        )
 
     @classmethod
     def from_dict(cls, d):
-        """Create from a dictionary. This class method will additionally
+        """Create from a dictionary. This class method will
         pre-process the passed dictionary to merge any field with an
         associated ``"_units"`` field into a :class:`pint.Quantity` container.
 
-        Since the class can be instantiated by either passing a file that
+        Sphere positions are specified in a text file, holding the x,y and z
+        coordinates for each sphere per line, separated by whitespace.
```

Comment:
Should this still be here?

Reply:
Yes, unless you think the sphere positions should be specified differently